### PR TITLE
Refactored Shelly 2.5 to use HttpBridge

### PR DESCRIPTION
**Pull request by Sn0w3y**

### CODE REVIEW

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java [hunk 1]
```diff
@@ -123,6 +168,164 @@ public default void setRelay1(boolean value) throws OpenemsNamedException {
 		this.getRelay1Channel().setNextWriteValue(value);
 	}
 
+	/**
+	 * Gets the Channel for {@link ChannelId#RELAY_1_OVERTEMP}.
+	 *
+	 * @return the Channel
+	 */
+	public default BooleanWriteChannel getRelay1OvertempChannel() {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java:176]: The purpose of those methods is to allow easy access of channel values in (complicated) code. If you don't use them anywhere, it's not required to add them.

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java [hunk 2]
```diff
@@ -41,6 +41,29 @@ public static enum ChannelId implements io.openems.edge.common.channel.ChannelId
 				.accessMode(AccessMode.READ_WRITE) //
 				.persistencePriority(PersistencePriority.HIGH) //
 				.onChannelSetNextWriteMirrorToDebugChannel(ChannelId.DEBUG_RELAY_1)),
+
+		/**
+		 * Indicates whether the associated meter is functioning properly.
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java:46]: Below its `is in Overtemp-State.` which sounds better suitable.

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 3]
```diff
@@ -55,14 +61,25 @@ public IoShelly25Impl() {
 	}
 
 	@Activate
-	private void activate(ComponentContext context, Config config) {
+	protected void activate(ComponentContext context, Config config) {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:64]: Why `protected`?

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 4]
```diff
@@ -75,20 +92,19 @@ public BooleanWriteChannel[] digitalOutputChannels() {
 	public String debugLog() {
 		var b = new StringBuilder();
 		var i = 1;
-		for (WriteChannel<Boolean> channel : this.digitalOutputChannels) {
+		for (BooleanWriteChannel channel : this.digitalOutputChannels) {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:95]: Use `var` for more modern coding style

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 5]
```diff
@@ -75,20 +92,19 @@ public BooleanWriteChannel[] digitalOutputChannels() {
...
 			String valueText;
 			var valueOpt = channel.value().asOptional();
 			if (valueOpt.isPresent()) {
-				valueText = valueOpt.get() ? "x" : "-";
+				valueText = valueOpt.get() ? "ON" : "OFF";
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:99]: You did not mention this change in the PR description. We use the notion of `x`, `-` and `?` in many places in OpenEMS. Why change it here?

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 6]
```diff
@@ -100,68 +116,86 @@ public void handleEvent(Event event) {
...
+				final var relay = JsonUtils.getAsJsonObject(relays.get(i));
+				final var relayIson = JsonUtils.getAsBoolean(relay, "ison");
+				final var relayOverpower = JsonUtils.getAsBoolean(relay, "overpower");
+				final var relayOvertemp = JsonUtils.getAsBoolean(relay, "overtemperature");
+
+				if (i == 0) {
+					relay1 = relayIson;
+					overtemp1 = relayOvertemp;
+					overpower1 = relayOverpower;
+				} else if (i == 1) {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:155]: This logic assumes, that there `relays.size() == 2`. There are multiple ways to do it better, e.g. 
- `for (var i : new int[] { 0, 1 }) {`
- `for (int i = 0; i < 2; i++) {`
- add a `record` with relay, overtemp and overpower and call a static `parseRelay()` method twice

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 7]
```diff
@@ -100,68 +116,86 @@ public void handleEvent(Event event) {
...
+		if (writeValue.isEmpty()) {
 			return;
 		}
 		if (Objects.equals(readValue, writeValue.get())) {
-			// read value = write value
 			return;
 		}
-		this.shellyApi.setRelayTurn(index, writeValue.get());
+		final String url = this.baseUrl + "/relay/" + index + "?turn=" + (writeValue.get() ? "on" : "off");
+		this.httpBridge.get(url).whenComplete((t, e) -> {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:195]: Should this not set `this._setSlaveCommunicationFailed(e != null);` like in the other Shelly implementations?

**Review by sfeilmeier — CHANGES_REQUESTED** [review]
Many topics appear over and over again in reviews. I believe we should eventually create an AbstractShelly class that combines all of the generic logic...

**Comment by Sn0w3y** [comment]
@sfeilmeier do you think it makes sense to sum up the meter Values and set _setActivePower(summedUpValues) for this Shelly ?

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 8]
```diff
@@ -99,68 +111,88 @@ public void handleEvent(Event event) {
...
+		} catch (OpenemsNamedException e) {
+			this.logDebug(this.log, e.getMessage());
 		}
-		this._setRelay1(relay1IsOn);
-		this._setRelay2(relay2IsOn);
+
+		this._setRelay1(relay1State.relayIsOn());
+		this._setRelay2(relay2State.relayIsOn());
+
+		this._setRelay1Overtemp(relay1State.overtemp());
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:162]: This potentially produces a NullPointerException

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 9]
```diff
@@ -99,68 +111,88 @@ public void handleEvent(Event event) {
...
+			final var relays = JsonUtils.getAsJsonArray(jsonElement, "relays");
+
+			for (int i = 0; i < 2; i++) {
+				if (i == 0) {
+					relay1State = parseRelay(JsonUtils.getAsJsonObject(relays.get(0)));
+				} else if (i == 1) {
+					relay2State = parseRelay(JsonUtils.getAsJsonObject(relays.get(1)));
+				}
+			}
 			this._setSlaveCommunicationFailed(false);
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:154]: This is redundant/illogical

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java [hunk 10]
```diff
@@ -99,68 +111,88 @@ public void handleEvent(Event event) {
...
+	private void processHttpResult(JsonElement result, Throwable error) {
+		this._setSlaveCommunicationFailed(result == null);
+		RelayState relay1State = null;
+		RelayState relay2State = null;
 
+		try {
+			JsonElement jsonElement = JsonUtils.getAsJsonElement(result);
+			final var relays = JsonUtils.getAsJsonArray(jsonElement, "relays");
+
+			for (int i = 0; i < 2; i++) {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25Impl.java:147]: This loop is not required

File: io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java [hunk 11]
```diff
@@ -123,6 +168,164 @@ public default void setRelay1(boolean value) throws OpenemsNamedException {
...
+	public default void setRelay1Overpower(boolean value) throws OpenemsNamedException {
+		this.getRelay1OverpowerChannel().setNextWriteValue(value);
+	}
+
+	/**
+	 * Gets the Channel for {@link ChannelId#RELAY_2_OVERPOWER}.
+	 *
+	 * @return the Channel
+	 */
+	public default BooleanWriteChannel getRelay2OverpowerChannel() {
```
Comment (sfeilmeier) [inline io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly25/IoShelly25.java:293]: This is not a WriteChannel.

**Review by sfeilmeier — APPROVED** [review]
Leider waren immer noch sehr viele Programmierfehler im Code. Ich muss die Qualitätsanforderungen hoch halten, deshalb kann ich das dann so nicht einfach mergen. So kommen wir mit den Shellys leider nur langsam voran. Von umfangreichen JUnit-Tests (z. B. parsing tests für die JSON-Response) spreche ich noch gar nicht.

Diesen PR habe ich jetzt nochmal selbst einigermaßen bereinigt. Bitte schau dir mal meine Commits und Kommentare an.

**Pull request merged**